apps/agreement/contracts/Agreement.sol

```diff
@@ -0,0 +1,1338 @@
+        IDisputable _disputable,
+        DisputableInfo storage _disputableInfo,
+        ERC20 _collateralToken,
+        uint256 _actionAmount,
+        uint256 _challengeAmount,
+        uint64 _challengeDuration
+    )
+        internal
+    {
+        require(isContract(address(_collateralToken)), ERROR_TOKEN_NOT_CONTRACT);
+
+        Staking staking = stakingFactory.getOrCreateInstance(_collateralToken);
+        uint256 id = _disputableInfo.nextCollateralRequirementsId++;
+        CollateralRequirement storage collateralRequirement = _disputableInfo.collateralRequirements[id];
+        collateralRequirement.token = _collateralToken;
+        collateralRequirement.staking = staking;
+        collateralRequirement.actionAmount = _actionAmount;
+        collateralRequirement.challengeAmount = _challengeAmount;
+        collateralRequirement.challengeDuration = _challengeDuration;
+
+        emit CollateralRequirementChanged(_disputable, id);
+    }
+
+    /**
+    * @dev Tell whether an address has permission to challenge actions on a specific disputable app
+    * @param _disputable Disputable app being queried
+    * @param _challenger Address of the challenger
+    * @return True if the challenger can be challenge actions on the disputable app, false otherwise
+    */
+    function _canPerformChallenge(IDisputable _disputable, address _challenger) internal view returns (bool) {
+        IKernel currentKernel = kernel();
+        if (currentKernel == IKernel(0)) {
+            return false;
+        }
+
+        // TODO: update with new ACL version: no need to pass challenger address by parameter
+        bytes memory params = ConversionHelpers.dangerouslyCastUintArrayToBytes(arr(_challenger));
+        return currentKernel.hasPermission(_challenger, address(_disputable), CHALLENGE_ROLE, params);
+    }
+
+    /**
+    * @dev Tell whether an action can be challenged
+    * @param _actionId Identification number of the action to be queried
+    * @param _action Action instance to be queried
+    * @return True if the action can be challenged, false otherwise
+    */
+    function _canChallenge(uint256 _actionId, Action storage _action) internal view returns (bool) {
+        return _canProceed(_actionId, _action) && _action.disputable.canChallenge(_action.disputableActionId);
+    }
+
+    /**
+    * @dev Tell whether an action can be closed
+    * @param _actionId Identification number of the action to be queried
+    * @param _action Action instance to be queried
+    * @return True if the action can be closed, false otherwise
+    */
+    function _canClose(uint256 _actionId, Action storage _action) internal view returns (bool) {
+        return _canProceed(_actionId, _action) && _action.disputable.canClose(_action.disputableActionId);
+    }
+
+    /**
+    * @dev Tell whether an action can proceed.
+    * @dev An action can proceed if it is not:
+    * @dev  - Closed
+    * @dev  - Currently challenged
+    * @dev  - Ruled as voided
```

Comment:
I think these 2 lines are wrong

Reply:
I think these are fine, actions can proceed if they were ruled in favor of the submitter or voided